// tstl/ranges/algorithm/heap.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
//================================================================ 
/** @module std.ranges */
//================================================================
var base = __importStar(require("../../algorithm/heap"));
var factory_1 = require("../../iterator/factory");
var comparators_1 = require("../../functional/comparators");
/* ---------------------------------------------------------
    PUSH & POP
--------------------------------------------------------- */
/**
 * Make a heap.
 *
 * @param range An iterable ranged container.
 * @param comp A binary function predicates *x* element would be placed before *y*. When returns `true`, then *x* precedes *y*. Default is {@link less}.
 */
function make_heap(range, comp) {
    if (comp === void 0) { comp = comparators_1.less; }
    return base.make_heap(factory_1.begin(range), factory_1.end(range), comp);
}
exports.make_heap = make_heap;
/**
 * Push an element into heap.
 *
 * @param range An iterable ranged container.
 * @param comp A binary function predicates *x* element would be placed before *y*. When returns `true`, then *x* precedes *y*. Default is {@link less}.
 */
function push_heap(range, comp) {
    if (comp === void 0) { comp = comparators_1.less; }
    return base.push_heap(factory_1.begin(range), factory_1.end(range), comp);
}
exports.push_heap = push_heap;
/**
 * Pop an element from heap.
 *
 * @param range An iterable ranged container.
 * @param comp A binary function predicates *x* element would be placed before *y*. When returns `true`, then *x* precedes *y*. Default is {@link less}.
 */
function pop_heap(range, comp) {
    if (comp === void 0) { comp = comparators_1.less; }
    return base.pop_heap(factory_1.begin(range), factory_1.end(range), comp);
}
exports.pop_heap = pop_heap;
/* ---------------------------------------------------------
    SORT
--------------------------------------------------------- */
/**
 * Test whether a range is heap.
 *
 * @param range An iterable ranged container.
 * @param comp A binary function predicates *x* element would be placed before *y*. When returns `true`, then *x* precedes *y*. Default is {@link less}.
 *
 * @return Whether the range is heap.
 */
function is_heap(range, comp) {
    if (comp === void 0) { comp = comparators_1.less; }
    return base.is_heap(factory_1.begin(range), factory_1.end(range), comp);
}
exports.is_heap = is_heap;
/**
 * Find the first element not in heap order.
 *
 * @param range An iterable ranged container.
 * @param comp A binary function predicates *x* element would be placed before *y*. When returns `true`, then *x* precedes *y*. Default is {@link less}.
 *
 * @return Iterator to the first element not in heap order.
 */
function is_heap_until(range, comp) {
    if (comp === void 0) { comp = comparators_1.less; }
    return base.is_heap_until(factory_1.begin(range), factory_1.end(range), comp);
}
exports.is_heap_until = is_heap_until;
/**
 * Sort elements of a heap.
 *
 * @param range An iterable ranged container.
 * @param comp A binary function predicates *x* element would be placed before *y*. When returns `true`, then *x* precedes *y*. Default is {@link less}.
 */
function sort_heap(range, comp) {
    if (comp === void 0) { comp = comparators_1.less; }
    return base.sort_heap(factory_1.begin(range), factory_1.end(range), comp);
}
exports.sort_heap = sort_heap;
